### PR TITLE
Add back to back exponential parameters in ENGIN-X parameter file

### LINKED ISSUE #29879: Load ENGIN-X refined back to back parameters from xml file
As a user I want the back-to-back exponential parameters from the refinement of a standard sample to be loaded into mantid to provide a good initial guess for A.B and S when fitting peaks in the focussed data.

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

Refined the GSAS/Fullprof coeficients for A,B,S as a function of d-spacing (input parameters for back to back exponential convoluted with a Gaussian) for the ENGIN-X instrument in a Parameter.xml file.

In testing it became apparent the parameter S was getting overridden when a peak was added (i.e. the refined result was not being used) - this is unexpected behaviour given instrument scientists are keeping the Parameter.xml up to date for other instruments as well. Methods were added to the IPeakFunction class and exposed to python (via the property handler and fit browser) such that when a new peak is added, the fwhm of a peak is only set if the parameter that relates to the width has not already set.  For fitting functions where the parameter that determines the width isn't specified (IPeakFunction method overridden in the function) the fwhm is always set.

**To test:**
(1) Load in the ENGIN-X data using the script below

```
from mantid.simpleapi import *
Load(Filename=r'\\isis.cclrc.ac.uk\inst$\ndxenginx\instrument\data\cycle_12_4\ENGINX00193749.nxs', 
     OutputWorkspace='193749')
```

(2) Change the log level to Debug
(3) Plot any spectrum, zoom in on a peak and press Fit
(4) Right-click on plot, Select peak type -> BackToBackExponential
(5) Add a peak, in the log you should see that where the parameters have been determined from the contents of the Parameter.xml file the parameter values post unit conversion are output. For example 
`The FitParameter S = 29.6166 after result-unit conversion`
(6) Check these numbers match those in the fit property browser
(7) Plot the guess and check is reasonably well initialised (contrast with master/nightly if in doubt)

Fixes #29879

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/widgets/common/src/PropertyHandler.cpp [hunk 1]
```diff
@@ -1379,6 +1379,28 @@ double PropertyHandler::fwhm() const {
   return 0;
 }
 
+std::string PropertyHandler::getWidthParameterName() const {
+  if (m_pf) {
+    return m_pf->getWidthParameterName();
+  }
+  return std::string();
```
Comment (StephenSmith25) [inline qt/widgets/common/src/PropertyHandler.cpp:1386]: As above we can return ""
```suggestion
  return ""
```

File: qt/widgets/common/src/PropertyHandler.cpp [hunk 2]
```diff
@@ -1379,6 +1379,28 @@ double PropertyHandler::fwhm() const {
...
+}
+
+std::string PropertyHandler::getCentreParameterName() const {
+  if (m_pf) {
+    return m_pf->getCentreParameterName();
+  }
+  return std::string();
```
Comment (StephenSmith25) [inline qt/widgets/common/src/PropertyHandler.cpp:1393]: As above we can return ""
```suggestion
  return ""
```

File: qt/widgets/common/inc/MantidQtWidgets/Common/PropertyHandler.h [hunk 3]
```diff
@@ -177,6 +177,9 @@ class EXPORT_OPT_MANTIDQT_COMMON PropertyHandler
   double height() const;
   double centre() const;
   double fwhm() const;
+  std::string getWidthParameterName() const;
+  std::string getCentreParameterName() const;
+  bool getParameterNameExplicitlySet(const std::string &param) const;
```
Comment (StephenSmith25) [inline qt/widgets/common/inc/MantidQtWidgets/Common/PropertyHandler.h:182]: Same as above: Maybe this should be called `isParameterNameExplicitlySet` or  `isParameterExplicitlySet`, the current name `getParameterNameExplicitlySet` makes it seem like its returning a parameter or parameter name.

File: qt/widgets/common/inc/MantidQtWidgets/Common/FitPropertyBrowser.h [hunk 4]
```diff
@@ -321,6 +321,10 @@ class EXPORT_OPT_MANTIDQT_COMMON FitPropertyBrowser
   double getPeakHeightOf(const QString &prefix);
   void setPeakFwhmOf(const QString &prefix, double value);
   double getPeakFwhmOf(const QString &prefix);
+  std::string getWidthParameterNameOf(const QString &prefix);
+  std::string getCentreParameterNameOf(const QString &prefix);
+  bool getParameterNameExplicitlySetOf(const QString &prefix,
```
Comment (StephenSmith25) [inline qt/widgets/common/inc/MantidQtWidgets/Common/FitPropertyBrowser.h:326]: Open for discussion, but maybe this should be called `isParameterNameExplicitlySet` or  `isParameterExplicitlySet`, the current name `getParameterNameExplicitlySet` makes it seem like its returning a parameter or parameter name.

**Review by StephenSmith25 — COMMENTED** [review]
This works well and is a great addition to stop these values from being overwritten after someone has gone to the effort to put the parameters in a file. Just some minor comments – and two queries:
-	Could this do with some release notes somewhere?
-	Would it be possible to add a unit test, possibly to the FitPropertyBrowser python tests, to make sure values are not being overwritten? Might not be possible, but would be nice to have more tests in this area.

File: Framework/API/inc/MantidAPI/IPeakFunction.h [hunk 5]
```diff
@@ -61,6 +61,15 @@ class MANTID_API_DLL IPeakFunction : public IFunctionWithLocation {
...
   std::string getCentreParameterName() const;
 
+  /// Get the name of the parameter that is changed when the
+  /// fwhm is changed. By default this returns an empty string (unless
+  /// overridden in the individual functions) as some functions change two
+  /// params when the fwhm is set and others don't have a width (delta
+  /// func).This is intended for the BackToBackExponential-based peaks where
+  /// the width parameter (S) can be set in the intsurment parameter file and
+  /// this needs to be checked when a peak is added.
+  virtual std::string getWidthParameterName() const { return std::string(); }
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply Framework/API/inc/MantidAPI/IPeakFunction.h:71]: I did this following the advice in
 https://stackoverflow.com/questions/26587110/returning-an-empty-string-efficient-way-in-c
but I'm happy to change.

**Review by StephenSmith25 — APPROVED** [review]
Looks good to go 👍

**Pull request merged**